plato/servers/base.py

```diff
@@ -333,7 +340,29 @@ async def periodic_task(self):
         # If we are operating in asynchronous mode, aggregate the model updates received so far.
         if hasattr(Config().server,
                    'synchronous') and not Config().server.synchronous:
-            if len(self.updates) > 0:
+
+            # What is the minimum number of clients that must have reported before aggregation
+            # takes place?
+            minimum_clients = 0
+            if hasattr(Config().server, 'minimum_clients_aggregated'):
+                minimum_clients = Config().server.minimum_clients_aggregated
+
+            # Is there any training clients who are currently training on models that are too
+            # `stale,` as defined by the staleness threshold?
+            staleness = 0
+            if hasattr(Config().server, 'staleness'):
+                staleness = Config().server.staleness
+
+            for __, client_data in self.training_clients.items():
+                if client_data['round'] < self.current_round - staleness:
+                    logging.info(
+                        "[Server #%d] Client %s is still working at round %s, which is "
+                        "beyond the staleness threshold %s compared to the current round %s. "
+                        "Nothing to process.", os.getpid(), client_data['id'],
+                        client_data['round'], staleness, self.current_round)
+                return
+
+            if len(self.updates) > minimum_clients:
```

Comment:
This logic needs to be reconsidered. 
1. ```self.current_round += 1```happens in ```select_client```, does it means the number of rounds equals the times of calling ```select_client``.
2. Assuming a client is very slow and other clients are fast, the system should be paused, while current implementation will continue the training once minimum_clients is reached. It does not follow the definition of SSP.
3. In #132, we in fact remove the usage of ```periodic_task```, which is driven by the timeout mechanism. The timeout mechanism is proposed in the PS-Lite paper (OSDI 2014). please distinguish it from SSP.

Reply:
I see no issues with the current logic. The current round advances whenever new clients are selected, because each of them is sent a newly aggregated model by the server. When a client is very slow, the aggregation process on the server will stop until this client catches up, as defined by 'staleness'.